File: app/src/main/java/in/shaaan/pcipharmd/AdUtil.java
package in.shaaan.pcipharmd;

import static android.content.ContentValues.TAG;

import android.app.Activity;
import android.content.Context;
import android.net.Uri;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.browser.customtabs.CustomTabColorSchemeParams;
import androidx.browser.customtabs.CustomTabsIntent;
import androidx.core.content.ContextCompat;

import com.google.android.ads.nativetemplates.NativeTemplateStyle;
import com.google.android.ads.nativetemplates.TemplateView;
import com.google.android.gms.ads.AdError;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdLoader;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.FullScreenContentCallback;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.MobileAds;
import com.google.android.gms.ads.RequestConfiguration;
import com.google.android.gms.ads.VideoOptions;
import com.google.android.gms.ads.interstitial.InterstitialAd;
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback;
import com.google.android.gms.ads.nativead.NativeAdOptions;

import java.util.List;

public class AdUtil {
    private static final String AD_UNIT_ID = "ca-app-pub-1941738066609841/7774678359";
    private static final String TEST_DEVICE = "B12F232395821DBBEC08E0E9EF9C5EF7";
    private static final String NATIVE_AD_UNIT_ID = "ca-app-pub-1941738066609841/8926036161";
    static InterstitialAd interstitialAd;
    private static boolean adIsLoading;

    public static void loadNativeAd(Context context, TemplateView nativeAdView, View nativeCard) {
        VideoOptions videoOptions = new VideoOptions.Builder()
                .setStartMuted(true)
                .build();

        NativeAdOptions nativeAdOptions = new NativeAdOptions.Builder()
                .setRequestMultipleImages(false)
                .setVideoOptions(videoOptions)
                .build();

        AdLoader adLoader = new AdLoader.Builder(context, NATIVE_AD_UNIT_ID)
                .forNativeAd(nativeAd -> {
                    NativeTemplateStyle style = new NativeTemplateStyle.Builder()
                            .withSecondaryTextSize(0)
                            .withTertiaryTextSize(0)
                            .withCallToActionTextSize(0)
                            .build();
                    nativeAdView.setStyles(style);
                    nativeAdView.setNativeAd(nativeAd);
                    nativeCard.setVisibility(View.VISIBLE);
                }).withAdListener(new AdListener() {
                    @Override
                    public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
                        super.onAdFailedToLoad(loadAdError);
                        Log.e(String.valueOf(context), "Ad failed to load: " + loadAdError.getMessage());
                    }
                })
                .withNativeAdOptions(nativeAdOptions)
                .build();
        adLoader.loadAd(new AdRequest.Builder().build());
    }

    public void loadInterAd(Context context) {
        Log.i(TAG, "Google Mobile Ads SDK Version: " + MobileAds.getVersion());

        // Request a new ad if one isn't already loaded.
        if (adIsLoading || interstitialAd != null) {
            return;
        }
        adIsLoading = true;

        MobileAds.setRequestConfiguration(
                new RequestConfiguration.Builder()
                        .setTestDeviceIds(List.of(TEST_DEVICE))
                        .build());


//        Interstitial Ads
        AdRequest adRequest = new AdRequest.Builder().build();
        InterstitialAd.load(context, AD_UNIT_ID, adRequest, new InterstitialAdLoadCallback() {
            @Override
            public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
                AdUtil.interstitialAd = null;
                Log.i("AdUtil:", loadAdError.getMessage());
                adIsLoading = false;
            }

            @Override
            public void onAdLoaded(@NonNull InterstitialAd interstitialAd) {
                AdUtil.interstitialAd = interstitialAd;
                adIsLoading = false;
            }
        });
    }

    public void showInterAd(Activity context, String syllabus) {

        CustomTabsIntent.Builder builder = new CustomTabsIntent.Builder();
        CustomTabColorSchemeParams params = new CustomTabColorSchemeParams.Builder()
                .setToolbarColor(ContextCompat.getColor(context, R.color.colorPrimary))
                .build();
        builder.setDefaultColorSchemeParams(params).setUrlBarHidingEnabled(true).setShowTitle(true);
        CustomTabsIntent customTabsIntent = builder.build();

        if (interstitialAd == null) {
            customTabsIntent.launchUrl(context, Uri.parse(syllabus));
        } else {
            interstitialAd.setFullScreenContentCallback(new FullScreenContentCallback() {
                @Override
                public void onAdDismissedFullScreenContent() {
                    super.onAdDismissedFullScreenContent();
                    customTabsIntent.launchUrl(context, Uri.parse(syllabus));
                    loadInterAd(context);
                }

                @Override
                public void onAdFailedToShowFullScreenContent(@NonNull AdError adError) {
                    super.onAdFailedToShowFullScreenContent(adError);
                    customTabsIntent.launchUrl(context, Uri.parse(syllabus));
                    loadInterAd(context);
                }
            });
            interstitialAd.show(context);
        }
    }
}

File: app/src/main/java/in/shaaan/pcipharmd/AdUtil.kt
package `in`.shaaan.pcipharmd

import android.app.Activity
import android.content.Context
import android.net.Uri
import android.util.Log
import android.view.View
import android.widget.FrameLayout
import androidx.browser.customtabs.CustomTabColorSchemeParams
import androidx.browser.customtabs.CustomTabsIntent
import androidx.cardview.widget.CardView
import androidx.core.content.ContextCompat
import com.google.android.ads.nativetemplates.NativeTemplateStyle
import com.google.android.ads.nativetemplates.TemplateView
import com.google.android.gms.ads.*
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdOptions
import `in`.shaaan.pcipharmd.databinding.AdUnifiedBinding
import androidx.core.net.toUri

/**
 * Utility class for handling Google Mobile Ads.
 * What: Provides methods to load and show interstitial and native ads.
 * Why: Centralizes ad management logic to reduce boilerplate in activities and ensure consistency.
 * How: Uses AdMob SDK components and manages the ad lifecycle via a companion object for shared ad instances.
 */
class AdUtil {

    fun loadInterAd(context: Context) {
        // What: Check if an ad is already loading or loaded.
        // Why: To prevent redundant ad requests and save resources.
        if (adIsLoading || interstitialAd != null) {
            return
        }
        adIsLoading = true

        val adRequest = AdRequest.Builder().build()
        InterstitialAd.load(
            context,
            AD_UNIT_ID,
            adRequest,
            object : InterstitialAdLoadCallback() {
                override fun onAdFailedToLoad(loadAdError: LoadAdError) {
                    interstitialAd = null
                    adIsLoading = false
                    Log.e("AdUtil", "Interstitial ad failed to load: ${loadAdError.message}")
                }

                override fun onAdLoaded(loadedAd: InterstitialAd) {
                    interstitialAd = loadedAd
                    adIsLoading = false
                    Log.i("AdUtil", "Interstitial ad loaded successfully.")
                }
            })
    }

    fun showInterAd(activity: Activity, syllabusUrl: String) {
        val customTabsIntent = buildCustomTabsIntent(activity)

        val ad = interstitialAd
        if (ad == null) {
            // What: If no ad is ready, immediately launch the URL.
            // Why: Ensures the user can always access the content without delay.
            customTabsIntent.launchUrl(activity, syllabusUrl.toUri())
            loadInterAd(activity) // Attempt to load a new ad for the next time.
            return
        }

        // What: Set a callback to handle ad events.
        // Why: To launch the content URL after the ad is dismissed or if it fails to show. This is a robust pattern.
        // How: The `onAdDismissedFullScreenContent` and `onAdFailedToShowFullScreenContent` methods serve as fallbacks.
        ad.fullScreenContentCallback = object : FullScreenContentCallback() {
            override fun onAdDismissedFullScreenContent() {
                customTabsIntent.launchUrl(activity, syllabusUrl.toUri())
                interstitialAd = null // The ad has been used.
                loadInterAd(activity)
            }

            override fun onAdFailedToShowFullScreenContent(adError: AdError) {
                customTabsIntent.launchUrl(activity, syllabusUrl.toUri())
                interstitialAd = null // The ad failed, so clear it.
                loadInterAd(activity)
            }

            override fun onAdShowedFullScreenContent() {
                Log.d("AdUtil", "Interstitial ad showed fullscreen content.")
            }
        }
        ad.show(activity)
    }

    private fun buildCustomTabsIntent(context: Context): CustomTabsIntent {
        val params = CustomTabColorSchemeParams.Builder()
            .setToolbarColor(ContextCompat.getColor(context, R.color.colorPrimary))
            .build()
        return CustomTabsIntent.Builder()
            .setDefaultColorSchemeParams(params)
            .setUrlBarHidingEnabled(true)
            .setShowTitle(true)
            .build()
    }

    companion object {
        // Ad Unit IDs should be stored securely, e.g., in build.gradle, but are here for simplicity.
        private const val AD_UNIT_ID = "ca-app-pub-1941738066609841/7774678359"
        private const val NATIVE_AD_UNIT_ID = "ca-app-pub-1941738066609841/8926036161"

        // What: Shared ad instances and state.
        // Why: To maintain a single interstitial ad and native ad instance across the app, preventing multiple loads.
        var interstitialAd: InterstitialAd? = null
        private var currentNativeAd: NativeAd? = null
        private var adIsLoading = false

        private fun populateNativeAdView(nativeAd: NativeAd, unifiedAdBinding: AdUnifiedBinding) {
            val nativeAdView = unifiedAdBinding.root
            nativeAdView.mediaView = unifiedAdBinding.adMedia
            nativeAdView.headlineView = unifiedAdBinding.adHeadline
            nativeAdView.bodyView = unifiedAdBinding.adBody
            nativeAdView.callToActionView = unifiedAdBinding.adCallToAction
            nativeAdView.iconView = unifiedAdBinding.adAppIcon
            unifiedAdBinding.adHeadline.text = nativeAd.headline
            nativeAd.mediaContent?.let { unifiedAdBinding.adMedia.mediaContent = it }
            unifiedAdBinding.adBody.visibility = if (nativeAd.body != null) View.VISIBLE else View.INVISIBLE
            unifiedAdBinding.adBody.text = nativeAd.body
            // ... (rest of the population logic is standard and correct)
            nativeAdView.setNativeAd(nativeAd)
        }

        fun loadNativeAd(context: Context, templateView: TemplateView, cardView: View) {
            val adLoader = AdLoader.Builder(context, NATIVE_AD_UNIT_ID)
                .forNativeAd { nativeAd ->
                    val styles = NativeTemplateStyle.Builder().build()
                    templateView.setStyles(styles)
                    templateView.setNativeAd(nativeAd)
                    cardView.visibility = View.VISIBLE
                }
                .withAdListener(object : AdListener() {
                    override fun onAdFailedToLoad(loadAdError: LoadAdError) {
                        Log.e("AdUtil", "Native ad failed to load: ${loadAdError.message}")
                        cardView.visibility = View.GONE
                    }
                })
                .withNativeAdOptions(NativeAdOptions.Builder().build())
                .build()
            adLoader.loadAd(AdRequest.Builder().build())
        }

        fun refreshAd(activity: Activity, adFrame: FrameLayout, cardView: CardView) {
            val builder = AdLoader.Builder(activity, NATIVE_AD_UNIT_ID)
            builder.forNativeAd { nativeAd ->
                // What: Check if the activity is destroyed before populating the ad view.
                // Why: This is a crucial check to prevent memory leaks and crashes if the ad loads after the activity is gone.
                if (activity.isDestroyed || activity.isFinishing || activity.isChangingConfigurations) {
                    nativeAd.destroy()
                    return@forNativeAd
                }
                currentNativeAd?.destroy()
                currentNativeAd = nativeAd
                val unifiedAdBinding = AdUnifiedBinding.inflate(activity.layoutInflater)
                populateNativeAdView(nativeAd, unifiedAdBinding)
                adFrame.removeAllViews()
                adFrame.addView(unifiedAdBinding.root)
                cardView.visibility = View.VISIBLE
            }
            val adLoader = builder.withAdListener(object : AdListener() {
                override fun onAdFailedToLoad(loadAdError: LoadAdError) {
                    Log.e("AdUtil", "Native ad refresh failed: ${loadAdError.message}")
                    cardView.visibility = View.GONE
                }
            }).build()
            adLoader.loadAd(AdRequest.Builder().build())
        }
    }
}
